src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -235,6 +238,57 @@ GcsPlacementGroupScheduler::GetOrConnectLeaseClient(const rpc::Address &raylet_a
   return iter->second;
 }
 
+void GcsPlacementGroupScheduler::OnPlacementGroupResourceReserveDone(
+    const std::shared_ptr<GcsPlacementGroup> &placement_group,
+    const std::vector<std::shared_ptr<BundleSpecification>> &bundles,
+    const std::shared_ptr<BundleLocation> &bundle_locations,
+    const std::function<void(std::shared_ptr<GcsPlacementGroup>)>
+        &schedule_failure_handler,
+    const std::function<void(std::shared_ptr<GcsPlacementGroup>)>
+        &schedule_success_handler) {
+  const auto &placement_group_id = placement_group->GetPlacementGroupID();
+  if (placement_group_leasing_in_progress_.find(placement_group_id) ==
+          placement_group_leasing_in_progress_.end() ||
+      bundle_locations->size() != bundles.size()) {
+    // If the lease request has been already cancelled
+    // or not every lease request succeeds,
+    for (auto &iter : *bundle_locations) {
+      auto &bundle_spec = iter.second.second;
+      auto &node_id = iter.second.first;
+      CancelResourceReserve(bundle_spec, gcs_node_manager_.GetNode(node_id));
+    }
+    schedule_failure_handler(placement_group);
+  } else {
+    // If we successfully created placement group, store them to GCS.
+    rpc::ScheduleData data;
+    for (const auto &iter : bundles) {
+      // TODO(ekl) this is a hack to get a string key for the proto
+      auto key = iter->PlacementGroupId().Hex() + "_" + std::to_string(iter->Index());
+      data.mutable_schedule_plan()->insert(
+          {key, (*bundle_locations)[iter->BundleId()].first.Binary()});
+    }
+    RAY_CHECK_OK(gcs_table_storage_->PlacementGroupScheduleTable().Put(
+        placement_group_id, data,
+        [schedule_success_handler, placement_group](Status status) {
+          schedule_success_handler(placement_group);
+        }));
+    // Update `node_to_leased_bundles_`.
+    for (const auto &iter : *bundle_locations) {
+      const auto &location = iter.second;
+      node_to_leased_bundles_[location.first].push_back(location.second);
```

Comment:
@ffbin This is not properly GC'ed currently. Do you have any idea how we can Gc it? It looks a little tricky.

Reply:
How about gc node_to_leased_bundles_ when we destroy placement group?

Reply:
That's possible. Let me add the logic.